### PR TITLE
Feat/add ts generics support

### PR DESCRIPTION
**Pull request by aadilhasan**
- Added generic type support

### CODE REVIEW

File: src/lib/apis/fetchPaginatedQuery.ts [hunk 1]
```diff
@@ -80,20 +80,20 @@ export async function fetchPaginatedQuery(
 
     if (paginationData.hasNextPage) {
       nextCursorsCache.push(paginationData.nextCursors);
-      const upatedQuery = await removeQueriesIfNoNextPage(
+      const updatedQuery = await removeQueriesIfNoNextPage(
```
Comment (aadilhasan) [inline src/lib/apis/fetchPaginatedQuery.ts:83]: fixed typo

File: src/lib/hooks/useQuery.ts [hunk 2]
```diff
@@ -30,22 +44,39 @@ export function useLazyQuery(
...
 
   const handleResponse = useCallback(
-    (res: Awaited<FetchQueryReturnType>) => {
-      if (!res) return;
-      const { data: rawData, error } = res;
-      originalData.current = rawData;
-      const data = rawData ? callbacksRef.current.dataFormatter(rawData) : null;
+    (res: Awaited<FetchQueryReturnType<ResponseType>>) => {
+      let data: ReturnType<Formatter> | null = null;
+      let error = null;
```
Comment (aadilhasan) [inline src/lib/hooks/useQuery.ts:55]: a bit of logic change please review carefully

File: src/lib/hooks/useQueryWithPagination.ts [hunk 3]
```diff
@@ -66,56 +85,63 @@ export function useLazyQueryWithPagination(
...
-        getPrevPage,
-        hasNextPage,
-        hasPrevPage,
-      } = res;
-      nextRef.current = getNextPage;
-      prevRef.current = getPrevPage;
-      originalData.current = rawData;
-      const data = rawData ? callbacksRef.current.dataFormatter(rawData) : null;
+    (res: null | Awaited<FetchPaginatedQueryReturnType<ResponseType>>) => {
+      let data: ReturnType<Formatter> | null = null;
```
Comment (aadilhasan) [inline src/lib/hooks/useQueryWithPagination.ts:89]: a bit of logic change please review carefully

File: src/lib/utils/fetcher.tsx [hunk 4]
```diff
@@ -25,8 +25,8 @@ export async function _fetch<ResponseType = any>(
     const json = await res.json();
     const data = json?.data;
     let error = null;
-    if (json.errors) {
-      error = json.errors;
+    if (json?.errors) {
+      error = json?.errors;
```
Comment (kaushalmeena) [inline src/lib/utils/fetcher.tsx:29]: don't need redundant null check in line 29, if it is already checked in line 28

**Review by kaushalmeena — APPROVED** [review]

**Comment by kaushalmeena** [comment]
Just wanted to confirm that we are not getting type inference in `dataFormatter` parameter, is it normal behaviour?

Screenshot 2023-10-11 at 6 52 13 PM

**Pull request merged**